examples/integrations/hf_trainer_api/det_callback.py

```diff
@@ -0,0 +1,348 @@
+        self.tokenizer = tokenizer
+        self.tokenizer_options = tokenizer_options
+        self.load_last_checkpoint(args)
+
+        self.last_metrics: typing.Dict[str, float] = {}
+
+        searcher_config = self._det.get_cluster_info().trial._config["searcher"]
+        self.searcher_metric = searcher_config["metric"]
+        self.searcher_unit = list(searcher_config["max_length"].keys())[0]
+        self.searcher_max_length = list(searcher_config["max_length"].values())[0]
+        self.searcher_ops = self.core_context.searcher.operations()
+        self.current_op = next(self.searcher_ops)
+
+        self._check_searcher_compatibility(args)
+
+    def on_log(
+        self,
+        args: TrainingArguments,
+        state: TrainerState,
+        control: TrainerControl,
+        logs=None,
+        **kwargs,
+    ):
+        if state.is_world_process_zero:
+            metrics, metric_type = self._get_metrics(logs)
+            if metric_type == TRAIN:
+                if (
+                    "train_step" not in self.last_metrics
+                    or self.last_metrics["eval_step"] != state.global_step
+                ):
+                    self.core_context.train.report_training_metrics(
+                        steps_completed=state.global_step, metrics=metrics
+                    )
+                    metrics["train_step"] = state.global_step
+
+            elif metric_type == EVAL:
+                if (
+                    "eval_step" not in self.last_metrics
+                    or self.last_metrics["eval_step"] != state.global_step
+                ):
+                    self.core_context.train.report_validation_metrics(
+                        steps_completed=state.global_step, metrics=metrics
+                    )
+                    metrics["eval_step"] = state.global_step
+                    self.last_metrics.update(metrics)
+            else:
+                logging.warning(f"Metrics not reported: metric type = {metric_type}.")
+
+            self.last_metrics.update(metrics)
+
+    def _get_metrics(self, logs: typing.Dict) -> typing.Tuple[typing.Dict, str]:
+        metrics = logs
+        metric_type = get_metric_type(logs)
+        if self.filter_metrics:
+            metrics = {}
+            for k, v in logs.items():
+                if any(m in k for m in self.filter_metrics) is True:
+                    metrics[k] = v
+
+        return metrics, metric_type
+
+    def on_save(
+        self,
+        args: TrainingArguments,
+        state: TrainerState,
+        control: TrainerControl,
+        **kwargs,
+    ):
+        info = self._det.get_cluster_info()
+        if info is None:
+            # TODO: modify to support local mode
+            logging.warning("ClusterInfo is None: not running in a task. Skip saving.")
+            return
+
+        local_path = os.path.join(args.output_dir, f"checkpoint-{state.global_step}")
+
+        storage_manager = self.core_context.checkpoint._storage_manager
```

Comment:
We can't publish an example that includes using internal APIs.  They are internal, so teaching users to use them is asking for trouble.

The `._storage_manager` is an internal attribute.

Can this example be written against purely public APIs?  If not, what additional APIs would need to exist for that to be possible?